clustering/ejb/WFLY-16345_SFSB_externalization_to_remote_Infinispan_server.adoc

```diff
@@ -0,0 +1,91 @@
+= @Stateful Jakarta Enterprise Bean instance externalization to remote Infinispan cluster
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently, WildFly only supports the use of embedded Infinispan for persisting/replicating @Stateful Jakarta Enterprise Bean instances.
```

Comment:
@pferraro I am not sure I fully understood the following:

> WildFly only supports the use of embedded Infinispan for persisting/replicating @Stateful Jakarta Enterprise Bean instances

That's because I was able to configure WildFly like:

```
embed-server --server-config=standalone-ha.xml

# JDG Nodes
/socket-binding-group=standard-sockets/remote-destination-outbound-socket-binding=remote-jdg-server1:add(host=127.0.0.1, port=11223)
/socket-binding-group=standard-sockets/remote-destination-outbound-socket-binding=remote-jdg-server2:add(host=127.0.0.1, port=11224)

# create the remote-cache-container
batch
/subsystem=infinispan/remote-cache-container=remote-infinispan-cluster:add(default-remote-cluster=jdg-server-cluster, statistics-enabled=true, properties={infinispan.client.hotrod.auth_username=admin, infinispan.client.hotrod.auth_password=pass.1234})
/subsystem=infinispan/remote-cache-container=remote-infinispan-cluster/remote-cluster=jdg-server-cluster:add(socket-bindings=[remote-jdg-server1,remote-jdg-server2])
run-batch

# add hotrod store to existing "distributed-cache=dist"
/subsystem=infinispan/cache-container=ejb/distributed-cache=dist/store=hotrod:add(remote-cache-container=remote-infinispan-cluster, fetch-state=false, preload=false, passivation=false, purge=false, shared=true)

# make "distributed-cache=dist" the default cache for EJB
/subsystem=infinispan/cache-container=ejb:write-attribute(name=default-cache, value=dist)
```

And I could see data being cached on the Infinispan cluster:

![image](https://github.com/user-attachments/assets/24e42d45-cdcb-4c03-a6eb-d6307bb4f726)

which looks like the SFSB is actually cached by:

```java
@WebServlet(name = "LocalEjbServlet", urlPatterns = { "/ejbservlet" })
public class LocalEjbServlet extends HttpServlet {

    @Inject
    private LocalStatefulSB bean;

    @Override
    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
        resp.setContentType("text/plain");
        int serial = bean.getSerialAndIncrement();
        resp.getWriter().print(serial);
    }

    @Override
    public String getServletInfo() {
        return "Servlet invoking Stateful Session Bean to store serial.";
    }
}
```

Reply:
SUMMARY: 

The goal of this RFE is to get rid **entirely** of any cache on the WF node that is involved in handling SFSB: we agreed to include this information in the AD ... the WF node would either create the session and the SFSB or retrieve the whole session data, for an existing session, from the remote Infinispan cluster;

Every node, clustered or not, would be somewhat "stateless": every time the node receives a request it would have to store/retrieve session data from the remote Infinispan cluster;

The "embedded Infinispan" refers to any `invalidation-cache`, `distributed-cache`, `replicated-cache`, "whatever" cache on the WF node;

Please note that is doesn't make sense to use a `distributed-cache` backed by a hotrod store (as in the example above) because they would serve the same purpose and we would do it twice! this is why we want to get rid of the `distributed-cache` element in this scenario...

The challenge that we need to solve is handling SFSB expiration: we have several entries involved with handling one (or more) SFSB on a remote Infinispan cluster:
- one entry to handle expiration
- one entry to actually store the serialized SFSB
When expiration happens:
1. the remote Infinispan cluster send the expiration message to **all** the connected WF nodes
2. each WF node leverages the **consistent hashing** mechanism to find out if it is the primary owner and hence in charge for handling that expiration message
3. if it is, it sends a "remove and get" message to the remote Infinispan cluster which actually removes the cache entry and retrieves the serialized SFSB
4. the WF node re-assembles the SFSB and run the code associated with SFSB expiration
5. the nodes that are not in charge for handling the expiration message, receive it regardless, they just have to make sure the primary owner actually processes the message: here is the challenge because this would involve a lot of network traffic

^^^ @pferraro can you validate the above?